### ISSUE TITLE
Recursor Lua preoutquery rcode "kill" does not work reliable

### ISSUE THREAD

**Issue by paddg** [PowerDNS/pdns#5478]
Program: Recursor

Lua preoutquery: dq.rcode = -3 sometimes does not work. Reproducible!" 

Operating system: SLES12-SP2
Software version: Recursor Git head (gf616ca6), self compiled

# Steps to reproduce

### debug.lua:
```
function preoutquery ( dq )
   if dq.qname:equal('example.nonsense') then
      dq.rcode = -3
      return true
   end
   return false
end
```

### test.sh:
```
#!/bin/sh
dig +cd  @127.0.0.1 example.nonsense A  | grep HEADER
dig +cd  @127.0.0.1 example.nonsense MX | grep HEADER
dig +cd  @127.0.0.1 example.nonsense A  | grep HEADER
sleep 7
dig +cd  @127.0.0.1 example.nonsense A  | grep HEADER
```

### Recursor command:
`pdns_recursor --lua-dns-script=debug.lua
`

Start`./test.sh`

### Expected behaviour:
```
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 39977
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 41008
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 58942
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 5628
```
### Actual behaviour:
```
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 39977
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 41008
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 58942
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 5628
```

**Comment by pieterlexis** [comment]
Could you provide us with a trace output of the Recursor during this test?

**Comment by paddg** [comment]
I've updated the Rcursor command line! The trace: https://p.6core.net/p/7BKxdBwwUEzRmXUOSAUooBKA

**Comment by pieterlexis** [comment]
Thanks!

A quick look seems to indicate that `preoutquery` is not called when hunting for zone cuts.

@rgacogne want me to look at the code or do you have an idea why this is?

**Comment by rgacogne** [comment]
@lieter the `preoutquery()` script only deny queries for `example.nonsense` and we get a `NXDOMAIN` for `nonsense.` from the root during the zone cuts hunt. It's happening after the initial `negcache` lookup so we still try to do an outgoing query after the zone cut, triggering the `preoutquery()` script. On the next try the `negcache` lookup correctly picks up the fact that the entire `nonsense.` domain has been denied by the root. Not sure how that should be fixed.

**Comment by omoerbeek** [comment]
Likely the draft #10057 will influence this.

**Comment by rgacogne** [comment]
Yes but I think qname minimization will still lead to the same issue? Perhaps disabling both `qname-minimization` and `nothing-below-nxdomain`[1] would help.

[1]: https://docs.powerdns.com/recursor/settings.html#nothing-below-nxdomain
